src/v/cluster/producer_state.cc

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "producer_state.h"
+
+#include "cluster/logger.h"
+#include "cluster/producer_state_manager.h"
+#include "vassert.h"
+
+namespace cluster {
+
+std::optional<request_ptr> requests::last_request() const {
+    if (!_inflight_requests.empty()) {
+        return _inflight_requests.back();
+    } else if (!_finished_requests.empty()) {
+        return _finished_requests.back();
+    }
+    return std::nullopt;
+}
+
+requests::request_result_t
+requests::new_request(seq_t first, seq_t last, model::term_id current) {
+    vlog(
+      clusterlog.trace,
+      "new request attempt, first: {}, last: {}, term: {}, producer: {}",
+      first,
+      last,
+      current,
+      *this);
+    // gc any inflight requests from old terms
+    while (!_inflight_requests.empty()
+           && _inflight_requests.front()->_term < current
+           && _inflight_requests.front()->in_progress()) {
+        // Here we know for sure the term change, these in flight requests
+        // are going to fail anyway, mark them so.
+        _inflight_requests.front()->set_value(errc::not_leader);
```

Comment:
`not_leader` means that the request wasn't attempted and 100% can't be replicated; but it doesn't seem to be the case there, if the leadership changed mid replication the replication state is unknown and it might be committed. we should return timeout or unknown server error